lib/xgu/xgu.h

```diff
@@ -0,0 +1,535 @@
+    p = push_float(p, parameter);
+    return p;
+}
+
+
+/* ========================= *
+ * BEGIN XGU FUNCTIONS BLOCK *
+ * ========================= */
+
+inline
+DWORD* xgu_begin(DWORD* p, XguPrimitiveType type) {
+
+    // Force people to use xgu_end instead
+    assert(type != NV097_SET_BEGIN_END_OP_END);
+
+    return push_command_parameter(p, NV097_SET_BEGIN_END, type);
+}
+
+inline
+DWORD* xgu_end(DWORD* p) {
+    return push_command_parameter(p, NV097_SET_BEGIN_END, NV097_SET_BEGIN_END_OP_END);
+}
+
+inline
+DWORD* xgu_set_viewport_offset(DWORD* p, float xw, float yh, float avgDepth) {
+    p = push_command(p, NV097_SET_VIEWPORT_OFFSET, 4);
+    p = push_float(p, xw);
+    p = push_float(p, yh);
+    p = push_float(p, avgDepth);
+    return push_float(p, (avgDepth == 0.0f ? 0.0f : 16777215.0));
+}
+
+inline
+DWORD* xgu_set_viewport_scale(DWORD* p, float w, float h, float znear, float zfar) {
+    p = push_command(p, NV097_SET_VIEWPORT_SCALE, 4);
+    p = push_float(p, w);
+    p = push_float(p, h);
+    p = push_float(p, ((zfar - znear / 2.0) * 16777214));
+    return push_float(p, 65535.0);
+}
+
+inline
+DWORD* xgu_set_clip_min(DWORD* p, float znear) {
+    return push_command_float(p, NV097_SET_CLIP_MIN, znear * 16777216.0);
+}
+
+inline
+DWORD* xgu_set_clip_max(DWORD* p, float zfar) {
+    return push_command_float(p, NV097_SET_CLIP_MAX, zfar * 16777216.0);
+}
+
+inline
+DWORD* xgu_set_zstencil_clear_value(DWORD* p, uint32_t depth, uint32_t stencil) {
+    return push_command_parameter(p, NV097_SET_ZSTENCIL_CLEAR_VALUE, (depth<<8)|stencil);
+}
+
+inline
+DWORD* xgu_set_color_clear_value(DWORD* p, DWORD color) {
+    return push_command_parameter(p, NV097_SET_COLOR_CLEAR_VALUE, color);
+}
+
+inline
+DWORD* xgu_clear_surface(DWORD* p, XguClearSurface flags) {
+    /*
+     * flags should probably be a combination of values -
+     * not sure if using an enum allows such magic and/or
+     * if it hinders out-of-bounds values.
+     */
+    return push_command_parameter(p, NV097_CLEAR_SURFACE, flags);
+}
+
+inline
+DWORD* xgu_set_clear_rect_horizontal(DWORD* p, uint32_t x1, uint32_t x2) {
+    return push_command_parameter(p, NV097_SET_CLEAR_RECT_HORIZONTAL,
+                                  ((x2-1)<<16)|x1);
+}
+
+inline
+DWORD* xgu_set_clear_rect_vertical(DWORD* p, uint32_t y1, uint32_t y2) {
+    return push_command_parameter(p, NV097_SET_CLEAR_RECT_VERTICAL,
+                                  ((y2-1)<<16)|y1);
+}
+
+inline
+DWORD* xgu_set_clear_rect(DWORD* p,
+                          uint32_t x1, uint32_t x2,
+                          uint32_t y1, uint32_t y2) {
+    p = push_command(p, NV097_SET_CLEAR_RECT_HORIZONTAL, 2);
+    p = push_parameter(p, ((x2-1)<<16|x1));
+    return push_parameter(p, ((y2-1)<<16|y1));
+}
+
+inline
+DWORD* xgu_set_object(DWORD* p, uint32_t instance) {
+    return push_command_parameter(p, NV097_SET_OBJECT, instance);
+}
+
+inline
+DWORD* xgu_set_texgen_s(DWORD* p, XguTexgen tg) {
```

Comment:
Needs variations for the other components: S,T,R,Q

Reply:
Agreed. Is it safe to assume that they'll want the same type of parameter?

Reply:
Probably, yes.
Comparing to https://www.khronos.org/registry/OpenGL-Refpages/gl2.1/xhtml/glTexGen.xml there are some errors in certain cases (probably those settings don't make a whole lot of sense). We can add asserts for those later if necessary.